SMLHelper/Options/ConfigFile.cs

```diff
@@ -0,0 +1,77 @@
+﻿
+namespace SMLHelper.V2.Options
+{
+    using System.IO;
+    using System.Reflection;
+    using SMLHelper.V2.Interfaces;
+
+    /// <summary>
+    /// A simple default implementation of an <see cref="IJsonFile"/>.
+    /// </summary>
+    /// <seealso cref="V2.Utility.JsonUtils.Load{T}(T, bool, Oculus.Newtonsoft.Json.JsonConverter[])"/>
+    /// <seealso cref="V2.Utility.JsonUtils.Save{T}(T, Oculus.Newtonsoft.Json.JsonConverter[])"/>
+    /// <example>
+    /// <code>
+    /// using System;
+    /// using SMLHelper.V2.Options;
+    /// using SMLHelper.V2.Utility;
+    /// using UnityEngine;
+    /// 
+    /// public class MyConfig : ConfigFile
+    /// {
+    ///     public KeyCode ActivationKey { get; set; } = KeyCode.Backspace;
+    /// }
+    /// 
+    /// public static class MyMod
+    /// {
+    ///     public static MyConfig Config = JsonUtils.Load&lt;MyConfig&gt;();
+    ///     public static void Initialize() {
+    ///         string activationKey = KeyCodeUtils.KeyCodeToString(Config.ActiviationKey);
+    ///         Console.WriteLine($"[MyMod] LOADED: ActivationKey = {activationKey}");
+    ///         // Will print "[MyMod] LOADED: ActivationKey = Backspace" in Player.log on first run,
+    ///         // and whatever value is have saved in the config.json for the "ActivationKey" 
+    ///         // property on subsequent runs.
+    ///     }
+    /// }
+    /// </code>
+    /// </example>
+    public abstract class ConfigFile : IJsonFile
+    {
+        /// <summary>
+        /// The file path at which the JSON file is accessible for reading and writing.
+        /// </summary>
+        public string JsonFilePath { get; private set; }
+
+        /// <summary>
+        /// Creates a new instance of <see cref="ConfigFile"/>.
+        /// </summary>
+        /// <param name="fileName">The name of the <see cref="ConfigFile"/>, "config" by default.</param>
+        /// <param name="subfolder">Optional subfolder for the <see cref="ConfigFile"/>.</param>
+        /// <example>
+        /// <code>
+        /// using SMLHelper.V2.Options;
+        /// using UnityEngine;
+        /// 
+        /// public static class MyConfig : ConfigFile
+        /// {
+        ///     public KeyCode ActivationKey { get; set; } = KeyCode.Escape;
+        ///     public MyConfig() : base("options", "Config Files") { }
+        ///     // The config file will be stored at the path "QMods\YourModName\Config Files\options.json"
+        /// }
+        /// </code>
+        /// </example>
+        protected ConfigFile(string fileName = "config", string subfolder = null)
+        {
+            var path = Path.GetDirectoryName(Assembly.GetCallingAssembly().Location);
+            if (string.IsNullOrEmpty(fileName))
+            {
+                fileName = "config";
+            }
+            JsonFilePath = Path.Combine(
+                path,
+                string.IsNullOrEmpty(subfolder) ? string.Empty : subfolder,
+                $"{fileName}.json"
```

Comment:
Good catch on the whitespace, I'll definitely add this to the docs.

Reply:
I've looked at the documentation for `Path.Combine` which you linked about for some extra information, and found this:

> The parameters are not parsed if they have white space. Therefore, if path2 includes white space (for example, " \file.txt "), the Combine method appends path2 to path1.

So it appears that you _can_ include white space, they merely won't be _parsed_ (to remove extra slashes etc.), but they'll still be _combined_.

I went on to test this by creating a class derived from `ConfigFile`:
```cs
    internal class Config : ConfigFile
    {
        public KeyCode OverrideKey { get; set; } = KeyCode.Escape;
        public AutoLoadMode AutoLoadMode { get; set; } = AutoLoadMode.MostRecentlySaved;
        public KeyCode ToggleAutoLoadMode { get; set; } = KeyCode.LeftControl;

        public Config() : base(subfolder: " Config Files ") { }
    }
```
And it created a file "config.json" at the path "QMods\MyMod\ Config Files\config.json". The trailing white space was omitted - which I assume was enforced by Windows (Explorer does the same thing) - but otherwise it worked as you might expect. As such, it would be misleading to suggest it cannot contain white space.